modules/examples/seed/client/modules/app/src/main/scala/example/seed/client/app/ClientBoot.scala

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2017-2019 47 Degrees, LLC. <http://www.47deg.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package example.seed.client.app
+
+import cats.effect._
+import cats.syntax.functor._
+import example.seed.client.common.models._
+import example.seed.client.process._
+import example.seed.config.ConfigService
+import fs2.Stream
+import io.chrisdavenport.log4cats.Logger
+import io.chrisdavenport.log4cats.slf4j.Slf4jLogger
+import pureconfig.generic.auto._
+
+abstract class ClientBoot[F[_]: ConcurrentEffect: ContextShift] {
+
+  def avroPeopleServiceClient(host: String, port: Int)(
+      implicit L: Logger[F]): Stream[F, AvroPeopleServiceClient[F]] =
+    AvroPeopleServiceClient.createClient(host, port, sslEnabled = false)
+
+  def protoPeopleServiceClient(host: String, port: Int)(
+      implicit L: Logger[F]): Stream[F, ProtoPeopleServiceClient[F]] =
+    ProtoPeopleServiceClient.createClient(host, port, sslEnabled = false)
+
+  def runProgram(args: List[String]): Stream[F, ExitCode] = {
+    def setupConfig: F[SeedClientConfig] =
+      ConfigService[F]
+        .serviceConfig[ClientConfig]
+        .map(client => SeedClientConfig(client, ClientParams.loadParams(client.name, args)))
+
+    for {
+      config   <- Stream.eval(setupConfig)
+      logger   <- Stream.eval(Slf4jLogger.fromName[F](config.client.name))
+      exitCode <- clientProgram(config)(logger)
+    } yield exitCode
+  }
+
+  def clientProgram(config: SeedClientConfig)(implicit L: Logger[F]): Stream[F, ExitCode]
```

Comment:
An alternative could make `ClientBoot` extends `IOApp`:

```scala
abstract class ClientBoot[F[_]: ConcurrentEffect: ContextShift] {

  def avroPeopleServiceClient(host: String, port: Int)(
      implicit L: Logger[F]): Stream[F, AvroPeopleServiceClient[F]] =
    AvroPeopleServiceClient.createClient(host, port, sslEnabled = false)

  def protoPeopleServiceClient(host: String, port: Int)(
      implicit L: Logger[F]): Stream[F, ProtoPeopleServiceClient[F]] =
    ProtoPeopleServiceClient.createClient(host, port, sslEnabled = false)

  def clientProgram(config: SeedClientConfig)(implicit L: Logger[F]): Stream[F, ExitCode]

  def run(args: List[String]): IO[ExitCode] = {
    def setupConfig: F[SeedClientConfig] =
      ConfigService[F]
        .serviceConfig[ClientConfig]
        .map(client => SeedClientConfig(client, ClientParams.loadParams(client.name, args)))

    (for {
      config   <- Stream.eval(setupConfig)
      logger   <- Stream.eval(Slf4jLogger.fromName[F](config.client.name))
      exitCode <- clientProgram(config)(logger)
    } yield exitCode).compile.toList.map(_.headOption.getOrElse(ExitCode.Error))
  }
}
```

So, the mains apps just need to:

```scala
class AvroClientProgram[F[_]: ConcurrentEffect: ContextShift] extends ClientBoot[F] {

  def clientProgram(config: SeedClientConfig)(implicit L: Logger[F]): Stream[F, ExitCode] = {
    for {
      peopleClient <- avroPeopleServiceClient(config.client.host, config.client.port)
      result       <- Stream.eval(peopleClient.getPerson(config.params.request))
    } yield result.fold(_ => ExitCode.Error, _ => ExitCode.Success)
  }
}
```

Reply:
To do this I will need an F on each method because the `run` method will need to have everything on IO, I prefer to leave it at it is but I can change to that if you want